File: ios/Packages/GemstoneServices/Sources/Stores/DeveloperStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import typealias Gemstone.BannerState
import protocol Gemstone.GemDeveloperStore
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneDeveloperStore: GemDeveloperStore, @unchecked Sendable {
    private let transactionStore: TransactionStore
    private let assetStore: AssetStore
    private let stakeStore: StakeStore
    private let bannerStore: BannerStore
    private let priceStore: PriceStore

    public init(
        transactionStore: TransactionStore,
        assetStore: AssetStore,
        stakeStore: StakeStore,
        bannerStore: BannerStore,
        priceStore: PriceStore,
    ) {
        self.transactionStore = transactionStore
        self.assetStore = assetStore
        self.stakeStore = stakeStore
        self.bannerStore = bannerStore
        self.priceStore = priceStore
    }

    public func clearTransactions() async throws {
        _ = try transactionStore.clear()
    }

    public func clearTokens() async throws {
        _ = try assetStore.clearTokens()
    }

    public func clearDelegations() async throws {
        _ = try stakeStore.clearDelegations()
    }

    public func clearValidators() async throws {
        _ = try stakeStore.clearValidators()
    }

    public func clearPrices() async throws {
        _ = try priceStore.clear()
    }

    public func clearBanners() async throws {
        _ = try bannerStore.clear()
    }

    public func updateBannerStates(from: Gemstone.BannerState, to: Gemstone.BannerState) async throws {
        _ = try bannerStore.updateStates(from: from.toPrimitives(), to: to.toPrimitives())
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stores/DeveloperStore.kt
package com.gemwallet.android.data.services.gemstone.stores

import com.gemwallet.android.data.service.store.database.AssetsDao
import com.gemwallet.android.data.service.store.database.BannersDao
import com.gemwallet.android.data.service.store.database.PricesDao
import com.gemwallet.android.data.service.store.database.StakeDao
import com.gemwallet.android.data.service.store.database.TransactionsDao
import com.gemwallet.android.ext.toPrimitives
import com.wallet.core.primitives.AssetType
import uniffi.gemstone.GemDeveloperStore

class GemstoneDeveloperStore(
    private val transactionsDao: TransactionsDao,
    private val assetsDao: AssetsDao,
    private val stakeDao: StakeDao,
    private val bannersDao: BannersDao,
    private val pricesDao: PricesDao,
) : GemDeveloperStore {

    override suspend fun clearTransactions() = transactionsDao.deleteAll()

    override suspend fun clearTokens() = assetsDao.deleteTokens(AssetType.NATIVE)

    override suspend fun clearDelegations() = stakeDao.deleteAllDelegations()

    override suspend fun clearValidators() = stakeDao.deleteAllValidators()

    override suspend fun clearPrices() = pricesDao.deleteAll()

    override suspend fun clearBanners() = bannersDao.deleteAll()

    override suspend fun updateBannerStates(from: uniffi.gemstone.BannerState, to: uniffi.gemstone.BannerState) =
        bannersDao.updateStates(from.toPrimitives(), to.toPrimitives())
}
